orttraining/orttraining/python/training/_ortmodule_utils.py

```diff
@@ -0,0 +1,103 @@
+    size_t cuda_caching_allocator_raw_alloc_address() {
+        return reinterpret_cast<size_t>(&c10::cuda::CUDACachingAllocator::raw_alloc);
+    }
+    size_t cuda_caching_allocator_raw_delete_address() {
+        return reinterpret_cast<size_t>(&c10::cuda::CUDACachingAllocator::raw_delete);
+    }
+    """
+
+    return load_inline(name='inline_extension', cpp_sources=[torch_cuda_allocator_addresses_cpp_source],
+                       functions=['cuda_caching_allocator_raw_alloc_address',
+                                  'cuda_caching_allocator_raw_delete_address'],
+                       verbose=verbosity < Verbosity.WARNING, with_cuda=True)
+
+def _check_same_device(device, argument_str, *args):
+    '''Check that all tensor arguments in *args reside on the same device as the input device'''
+
+    for arg in args:
+        if arg is not None and isinstance(arg, torch.Tensor):
+            arg_device = torch.device(arg.device)
+            if arg_device != device:
+                raise RuntimeError(
+                    f"{argument_str} found on device {arg_device}, but expected it to be on module device {device}.")
+
+def _create_iobinding(io_binding, inputs, model, device):
+    '''Creates IO binding for a `model` inputs and output'''
+    for idx, value_info in enumerate(model.graph.input):
+        io_binding.bind_ortvalue_input(
+            value_info.name, _ortvalue_from_torch_tensor(inputs[idx]))
+
+    for value_info in model.graph.output:
+        io_binding.bind_output(value_info.name, device.type,
+                               device_id=_utils.get_device_index(device))
+
+def _run_forward(execution_session, onnx_model, device, *inputs, **kwargs):
+    """Runs the forward graph on execution_session with given model inputs and device"""
+
+    # Assert that the input and model device match
+    _check_same_device(
+        device, "Input argument to forward", *inputs)
+
+    # TODO: Try to reuse the output buffers as some of the output tensors are same sizes,
+    #   especially the backward graph outputs.
+    io_binding = execution_session.io_binding()
+    run_options = C.RunOptions()
+
+    # Use IO binding
+    _create_iobinding(io_binding, inputs, onnx_model, device)
+
+    # Run and return module outputs.
+    ort_output = execution_session.run_forward(io_binding, run_options)
+    forward_outputs, run_id = ort_output.ortvalues, ort_output.run_id
+
+    user_outputs = tuple(_ortvalue_to_torch_tensor(
+        forward_output) for forward_output in forward_outputs)
+
+    # Assert that the outputs and model device match
+    _check_same_device(
+        device, "Output argument from forward", *user_outputs)
+
+    output_info = [(output.shape, output.device, output.dtype) for output in user_outputs]
+    run_info = onnxruntime.training.RunStateInfo(run_id, run_options, io_binding, output_info)
+
+    # Return user outputs and forward run information
+    return user_outputs, run_info
```

Comment:
```suggestion
    return user_outputs, run_info

```

Reply:
Moved the _run_forward method to GraphExecutionManager.